atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/impl/StaticTableMappingService.java

```diff
@@ -39,21 +41,24 @@ public InitializingWrapper(StaticTableMappingService tableMappingService) {
 
         @Override
         public TableMappingService delegate() {
-            return tableMappingService;
+            if (tableMappingService.isInitialized()) {
+                return tableMappingService;
+            }
+            throw new NotInitializedException("TableMappingService");
         }
     }
 
     private volatile boolean isInitialized = false;
     private final KeyValueService kv;
 
-    public static StaticTableMappingService create(KeyValueService kv) {
-        return create(kv, true);
+    public static TableMappingService create(KeyValueService kv) {
+        return create(kv, AtlasDbConstants.DEFAULT_INITIALIZE_ASYNC);
     }
 
-    public static StaticTableMappingService create(KeyValueService kv, boolean initalizeAsync) {
-        StaticTableMappingService ret = new StaticTableMappingService(kv);
-        ret.initialize(initalizeAsync);
-        return ret;
+    public static TableMappingService create(KeyValueService kv, boolean initalizeAsync) {
+        StaticTableMappingService tableMappingService = new StaticTableMappingService(kv);
+        tableMappingService.initialize(initalizeAsync);
+        return tableMappingService.isInitialized() ? tableMappingService : new InitializingWrapper(tableMappingService);
```

Comment:
Hmm, this pattern occurs quite a bit. I'm thinking we can extract this pattern somewhere:

```java
    public static <T extends AsyncInitializer> T createMaybeAsync(
            boolean initializeAsync,
            Supplier<T> creator) {
        T object = creator.get();
        object.initialize(initializeAsync);
        return object.isInitialized() ? object : new InitializingWrapper(object);
    }
```

Not sure if it's worth the effort though.

Reply:
It's a bit tricky with the InitializingWrapper since it's a different one for each T so it gets nasty

Reply:
As discussed, we would need a Function<T, W> to send the wrapper the right object. Probably overkill for now.